value-set domain for variable-sensitivity-domain

owen-mc-diffblue opened the pull request:
Note this PR is not to `develop`, it is to `variable-sensitivity-domain`.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

owen-mc-diffblue committed:
Implement value sets

as a new abstract object.

owen-mc-diffblue committed:
Extend the interface with hashing and equality  for unordered-sets.

Commit:
If there is no context to wrap - don't wrap the objects

There is no value in wrapping an object with no additional information,
and there is currently some problems with wrapping

Commit:
Only enable location context if value set is turned off

There is currently some inconsistency between how the value set and the
interval domains handle wrapping. Specifically, the value set returns an
unwrapped object when eval'd, whereas the interval domain returns a
wrapped object. Since the rest of the code is compatible with the
interval suggestion, we temporarily disable wrapping if using value set

Commit:
Produce an error if value set is used with a context tracking argument

Commit:
Remove unusued boolean flag

The variable sensitivity domain is always on, if we're using the
variable sensitivity domain object factory

Commit:
Remove the way to set configuration options using only optionst

Commit:
Introduce some sensible config defaults

Commit:
Use the config to simplify the code

Commit:
Add tests for value-set domain

owen-mc-diffblue committed:
Add value-set flag to goto-analyzer

owen-mc-diffblue committed:
Make value set entries human-readable

Use from_expr instead of constant_expr::get_value() so we don't print hex values

owen-mc-diffblue committed:
Various miscellaneous fixes

Most of these are needed to make clang-format happy, or
the linter, or to make it compile on Windows

owen-mc-diffblue committed:
Revert "Update janalyzer Makefile for renamed reachable_instructions.cpp"

This reverts commit cac98cef

hannes-steffenhagen-diffblue commented on src/analyses/variable-sensitivity/abstract_object.h:
```diff
@@ -224,6 +224,16 @@ class abstract_objectt:public std::enable_shared_from_this<abstract_objectt>
     const abstract_object_visitort &visitor) const
   { return shared_from_this(); }
 
+  virtual size_t internal_hash() const
```
Where are these used?

owen-mc-diffblue replied:
They are used in `abstract_hashert` and `abstract_equalert`, further down in this file, which are in turn used for the definition of `abstract_object_sett` at the top of value_set_abstract_object.h, which is in the next commit.

owen-mc-diffblue committed:
Delete write_stack unit test

It was written in 2017 but was #if 0-ed out, so it needs to be updated to compile now.
ADA-496 has been filed to fix this test up and reintroduce it.

hannes-steffenhagen-diffblue commented on regression/goto-analyzer/value-set-function-pointers-arrays/test.desc:
```diff
@@ -0,0 +1,13 @@
+^file main.c line 29 function main: replacing function pointer by 2 possible targets$
+^file main.c line 40 function main: replacing function pointer by 2 possible targets$
+^main::1::fun1 \(\) -> value-set-begin: ptr ->\(f\), ptr ->\(g\), :value-set-end$
+^main::1::fun_array3 \(\) -> \{\[0\] = value-set-begin: ptr ->\(f\), ptr ->\(g\), :value-set-end$
```
where are `fun_array1` and 2?

owen-mc-diffblue replied:
`fun_array1` is read from when assigning a value to `fun1`. `fun_array2` is read from when assigning a value to `fun_array3`. There is little benefit to asserting what values they contain.

hannes-steffenhagen-diffblue commented on src/analyses/variable-sensitivity/variable_sensitivity_object_factory.h:
```diff
@@ -125,9 +125,19 @@ abstract_object_pointert variable_sensitivity_object_factoryt::
-    return initialize_context_abstract_object<
-      abstract_object_classt, context_abstract_objectt>(
-        type, top, bottom, e, enviroment, ns);
+  {
```
The message in this commit is confusing. Top/bottom are unrelated to context afaict.

owen-mc-diffblue replied:
The point is that the commit makes it return an unwrapped object. It just so happens that this is done differently for top and bottom.

hannes-steffenhagen-diffblue commented on src/analyses/variable-sensitivity/value_set_abstract_object.h:
```diff
@@ -0,0 +1,228 @@
+{
+public:
+  using abstract_object_sett = std::unordered_set<
+    abstract_object_pointert,
```
I believe this is incorrect? This should just be `exprt`.

NlightNFotis replied:
Yes, I believe so too.

owen-mc-diffblue replied:
I don't know the ins and outs, but that would be a major change, because `abstract_object_pointert` doesn't inherit from `exprt`

hannes-steffenhagen-diffblue commented on src/analyses/variable-sensitivity/interval_array_abstract_object.cpp:
```diff
@@ -38,6 +39,19 @@ static constant_interval_exprt eval_and_get_as_interval(
   const namespacet &ns)
 {
   auto evaluated_index = environment.eval(expr, ns);
+  auto maybe_interval = evaluated_index->unwrap_context();
```
I'm not really sure why we're changing interval array abstract object here? This should be unrelated to value set abstract objects.

owen-mc-diffblue replied:
I don't know because I didn't write the code. All the tests pass without this chunk of code, so it seems you're probably right.

hannes-steffenhagen-diffblue commented on src/analyses/variable-sensitivity/value_set_abstract_object.cpp:
```diff
@@ -0,0 +1,305 @@
+  std::size_t num_operands = expr.operands().size();
+  PRECONDITION(operands.size() == num_operands);
+
+  std::vector<abstract_object_sett> collective_operands;
```
These should just be `value_set_abstract_objectt`

owen-mc-diffblue replied:
Because of the recursive nature of `apply_comb`, we need to be able to access the `n`th entry of `collective_operands`, and hence it can't just be a set.

hannes-steffenhagen-diffblue commented on src/analyses/variable-sensitivity/value_set_abstract_object.cpp:
```diff
@@ -0,0 +1,305 @@
+  }
+
+  abstract_object_sett resulting_objects;
+  for_each_comb(
```
This bit needs to account for the possibility of an operand being top (or bottom I guess, although in that case I don't think we'd even evaluate this?).

owen-mc-diffblue replied:
It makes sense to me that that this should cope with top, but I have no idea how to implement it

NlightNFotis reviewed (approved):
Overall, it looks good to me.

The only thing I would ask for is a squash of the small inconsequential commits (like copyright changes, or removal of files or the tests being together in one commit if possible).

hannes-steffenhagen-diffblue commented:
We'll be taking this over
